feat: migrate package to TypeScript and publish types

michaelfaith opened the pull request:
This change migrates all source (runtime code and tests) to TypeScript, and as a result now results in type definitions being published with the package.

Because I knew this was going to be a big PR, I tried to break things up the best I could by commit.  I first `git mv` all runtime files, and then in separate commits, migrated each rule.  I then did the same thing for tests.

The actual changes to test files, beyond the file name change, was fairly minimal.  Most of them were just simple file name changes.  Only a few needed some additional work (e.g. types added to utility functions).  So, I feel pretty good about maintaining functionality.  

I will say, the type in `utils` aren't the best, and I had to resort to adding `@ts-expect-error` in a couple places, but even then, they're in a pretty decent state.  I wanted to keep the flow of the functions as close to their original form as possible in this initial cut over.  All of that is to say, there's opportunity to iterate on these internal types, to improve them further.  Of course, those changes wouldn't be breaking and could potentially come after v7 is released.  

In addition to all the runtime and test file changes, I addded a build script (using `tsup`) and updated the `package.json` with new entry point paths and to include `dist` instead of `lib`.  I also updated the publish GH workflow to run a build before `npm publish`.  

Supporting work that I landed prior to this:
- migrate `eslint.config.js` to typescript so we could use the plugin directly from source: https://github.com/eslint-community/eslint-plugin-eslint-plugin/pull/531
- Remove unneeded context compatibility functions: https://github.com/eslint-community/eslint-plugin-eslint-plugin/pull/532
- migrate the `eslint-remote-tester` config to be typescript so we could use the plugin directly from source: https://github.com/eslint-community/eslint-plugin-eslint-plugin/pull/533

Closes #310

michaelfaith commented:
If there are any other slices of work you think should be peeled off and broken out as a separate change, let me know.  I _think_ everything in here, for the most part, need to go together, but open for input on that.

bmish commented on lib/rules/no-unused-message-ids.ts:
```diff
@@ -99,21 +101,27 @@ const rule = {
               messageId.type === 'Literal'
                 ? [messageId]
-                : findPossibleVariableValues(messageId, scopeManager);
+                : findPossibleVariableValues(
+                    messageId as Identifier,
+                    scopeManager,
+                  );
```
Would prefer type checks over casting, and just bail out if it's not the right type.

```suggestion
                : typeof messageId === 'Identifier' ? findPossibleVariableValues(
                    messageId,
                    scopeManager,
                  ): [];
```

michaelfaith replied:
Updated

bmish commented on lib/rules/no-unused-message-ids.ts:
```diff
@@ -127,18 +135,23 @@ const rule = {
             values.length === 0 ||
             values.some((val) => val.type !== 'Literal') ||
-            isVariableFromParameter(node.value, scopeManager)
+            isVariableFromParameter(node.value as Identifier, scopeManager)
```
Avoid casting.

bmish commented on lib/rules/no-unused-message-ids.ts:
```diff
@@ -127,18 +135,23 @@ const rule = {
           }
 
-          values.forEach((val) => messageIdsUsed.add(val.value));
+          values.forEach(
+            (val) => 'value' in val && messageIdsUsed.add(val.value as string),
+          );
```
Avoid casting.

michaelfaith replied:
Updated to 
```ts
values
            .filter((val) => val.type === 'Literal')
            .map((val) => val.value)
            .filter((val) => typeof val === 'string')
            .forEach((val) => messageIdsUsed.add(val));
```

bmish commented on lib/rules/no-useless-token-range.ts:
```diff
@@ -51,33 +58,35 @@ const rule = {
         (arg.properties.length === 1 &&
           (getKeyName(arg.properties[0]) !== 'includeComments' ||
-            arg.properties[0].value.type !== 'Literal'))
+            (arg.properties[0] as Property).value.type !== 'Literal'))
```
type check instead of casting

michaelfaith replied:
Updated

bmish commented on lib/rules/no-useless-token-range.ts:
```diff
@@ -51,33 +58,35 @@ const rule = {
       );
     }
 
+    function isMemberExpression(node: Node): node is MemberExpression {
```
I believe we can avoid simple type-checking functions like this. Inlining the type check should work fine for narrowing the type.

michaelfaith replied:
Updated.  This was just for convenience, since there were so many.

bmish commented on lib/rules/no-useless-token-range.ts:
```diff
@@ -139,20 +150,22 @@ const rule = {
+                identifier.parent.parent.range![0],
+              ) +
+              sourceCode.getText(
+                (identifier.parent.parent as CallExpression).arguments[0],
```
Type check instead of casting.

michaelfaith replied:
This is another example of where we're already doing the check in the filter, above, it's just not narrowing that prop downstream, so it feels redundant.  But I'll go ahead and change it. 

<img width="506" height="169" alt="image" src="https://github.com/user-attachments/assets/78a3ae23-4461-4cf0-9cb9-7e2debdd4e90" />

michaelfaith replied:
Updated

bmish commented on lib/rules/no-useless-token-range.ts:
```diff
@@ -161,7 +174,9 @@ const rule = {
                   identifier.parent.parent,
-                  sourceCode.getText(identifier.parent.parent.arguments[0]),
+                  sourceCode.getText(
+                    (identifier.parent.parent as CallExpression).arguments[0],
```
Type check instead of casting.

bmish commented on lib/rules/report-message-format.ts:
```diff
@@ -41,23 +40,25 @@ const rule = {
-          !pattern.test(message.quasis[0].value.cooked)) ||
-        (staticValue && !pattern.test(staticValue.value))
+          !pattern.test(message.quasis[0].value.cooked ?? '')) ||
+        (staticValue && !pattern.test(staticValue.value as string))
```
Type check instead of casting.

bmish commented on lib/rules/require-meta-default-options.ts:
```diff
@@ -68,7 +68,7 @@ const rule = {
     if (!metaDefaultOptions) {
       context.report({
-        node: metaNode,
+        node: metaNode!,
```
Type check instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/rules/require-meta-docs-url.ts:
```diff
@@ -102,7 +102,7 @@ const rule = {
         }
 
-        if (isExpectedUrl(staticValue && staticValue.value)) {
+        if (isExpectedUrl(staticValue && (staticValue.value as string))) {
```
Type-checking instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/rules/require-meta-fixable.ts:
```diff
@@ -99,7 +98,9 @@ const rule = {
           if (
-            !['code', 'whitespace', null, undefined].includes(staticValue.value)
+            !['code', 'whitespace', null, undefined].includes(
+              staticValue.value as string,
```
Type-checking instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/rules/require-meta-fixable.ts:
```diff
@@ -111,7 +112,7 @@ const rule = {
           if (
             usesFixFunctions &&
-            !['code', 'whitespace'].includes(staticValue.value)
+            !['code', 'whitespace'].includes(staticValue.value as string)
```
Type-checking instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/rules/require-meta-fixable.ts:
```diff
@@ -121,7 +122,7 @@ const rule = {
             catchNoFixerButFixableProperty &&
             !usesFixFunctions &&
-            ['code', 'whitespace'].includes(staticValue.value)
+            ['code', 'whitespace'].includes(staticValue.value as string)
```
Type-checking instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/rules/require-meta-type.ts:
```diff
@@ -68,7 +62,7 @@ const rule = {
         }
 
-        if (!VALID_TYPES.has(staticValue.value)) {
+        if (!VALID_TYPES.has(staticValue.value as string)) {
```
Type-checking instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/utils.ts:
```diff
@@ -116,14 +161,19 @@ function isTypeScriptRuleHelper(node) {
       // export default rule;
       case 'ExportDefaultDeclaration': {
-        possibleNodes.push(statement.declaration);
+        possibleNodes.push(statement.declaration as Identifier);
```
Type-check instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/utils.ts:
```diff
@@ -141,23 +191,23 @@ function getRuleExportsESM(ast, scopeManager) {
             statement.declaration.type === 'VariableDeclaration'
               ? statement.declaration.declarations.map(
-                  (declarator) => declarator.init,
+                  (declarator) => declarator.init!,
```
Instead of forcing this to exist, can we filter out null/undefined afterward with `.filter(val=>!!val)`?

michaelfaith replied:
Updated

bmish commented on lib/utils.ts:
```diff
@@ -196,34 +246,37 @@ function getRuleExportsESM(ast, scopeManager) {
-
-    .reduce((currentExports, node) => {
+    .reduce<PartialRuleInfo>((currentExports, node) => {
+      const leftExpression = node.left as MemberExpression;
```
Type-check instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/utils.ts:
```diff
@@ -319,15 +381,15 @@ function findVariableValue(node, scopeManager) {
  * Ex: foo ? [a, b, c] : [d, e, f] will return [a, b, c, d, e, f]
- * @param {Node} node
- * @returns {Node[]} the list of elements
+ * @param node
```
```suggestion
```

bmish commented on lib/utils.ts:
```diff
@@ -399,19 +468,25 @@ export function getContextIdentifiers(scopeManager, ast) {
-      .find((variable) => variable.name === ruleInfo.create.params[0].name)
+      .find(
+        (variable) =>
+          variable.name === (ruleInfo.create.params[0] as Identifier).name,
```
Type-check instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/utils.ts:
```diff
@@ -420,7 +495,7 @@ export function getKeyName(property, scope) {
       if (scope) {
         const staticValue = getStaticValue(property.key, scope);
-        return staticValue ? staticValue.value : null;
+        return staticValue ? (staticValue.value as string) : null;
```
Type-check instead of casting.

michaelfaith replied:
Updated

bmish commented on lib/utils.ts:
```diff
@@ -861,47 +983,61 @@ export function getMessagesNode(ruleInfo, scopeManager) {
     : undefined;
 }
 
+const isProperty = (node: Node): node is Property => node.type === 'Property';
```
I don't think having this helper is necessary?

michaelfaith replied:
Updated

bmish commented on package.json:
```diff
@@ -3,28 +3,30 @@
     "release": "release-it",
     "test": "vitest run --coverage",
-    "test:remote": "eslint-remote-tester -c ./eslint-remote-tester.config.ts",
+    "test:remote": "eslint-remote-tester",
```
Does it detect the TS config file automatically now?

michaelfaith replied:
It does!

bmish commented on tests/lib/rules/require-meta-type.ts:
```diff
@@ -66,16 +66,6 @@ ruleTester.run('require-meta-type', rule, {
         create(context) {}
       };
     `,
-    {
-      code: `
-        const create = {};
-        module.exports = {
-          meta: {},
-          create,
-        };
-      `,
-      errors: [{ messageId: 'missing' }],
-    },
```
What was this removed for?

michaelfaith replied:
This was in the `valid` array, but had an `errors` property, which isn't correct.  Should I add it to the invalid array?

bmish replied:
I'm not sure if it's valid or invalid, but whichever one it is, it would be good to put it there.

michaelfaith replied:
Cool.  I'll add it to the `invalid` array.

michaelfaith replied:
Oh, actually, nevermind.  Its `create` doesn't have a `context`, so it's not detecting it as a rule.  I.e. `valid`.  I'll add it back to valid and remove the `errors`.

michaelfaith replied:
Done

bmish commented on types/estree.d.ts:
What's the purpose of this file?

bmish replied:
Can you add this as a comment to the file explaining what these types are for and where they come from?

michaelfaith commented on lib/rules/fixer-return.ts:
```diff
@@ -70,10 +83,9 @@ const rule = {
      * A fix is an object created by fixer.replaceText() for example and returned by the fix function.
-     * @param {ASTNode} node - node to check
-     * @returns {boolean}
+     * @param node - node to check
```
Updated.

bmish commented on lib/rules/meta-property-ordering.ts:
```diff
@@ -52,24 +54,29 @@ const rule = {
+    const order: string[] = context.options[0] || defaultOrder;
 
-    const orderMap = new Map(order.map((name, i) => [name, i]));
+    const orderMap = new Map<string | null, number>(
```
Oh nice! Did not realize that. Thanks.

michaelfaith commented on lib/rules/no-identical-tests.ts:
```diff
@@ -27,20 +27,13 @@ const rule = {
      * Create a unique cache key
-     * @param {object} test
-     * @returns {string}
+     * @param test
```
Updated

michaelfaith commented on lib/rules/no-meta-schema-default.ts:
```diff
@@ -43,31 +43,32 @@ const rule = {
         }
 
-        switch (key.name ?? key.value) {
+        switch ('name' in key ? key.name : 'value' in key ? key.value : '') {
```
That sufficiently narrows the true condition, but not `value`.  I'd still need something more to use `key.value`

michaelfaith commented on lib/rules/no-missing-placeholders.ts:
```diff
@@ -96,9 +96,9 @@ const rule = {
             data,
           } of reportMessagesAndDataArray.filter((obj) => obj.message)) {
-            const messageStaticValue = getStaticValue(message, scope);
+            const messageStaticValue = getStaticValue(message!, scope);
```
the filter doesn't do enough to narrow the type here, unfortunately.  Since the filter ensures `message` has a value, I thought it'd be ok assert here.  There shouldn't ever be a case where `message` is falsey, but I can add the condition.

bmish replied:
Yeah generally would prefer to use an early return even if not technically necessary. This helps avoid issues that casting can hide.

However, it looks like an inline type guard could also help here:

```ts
arr
  .filter((val): val is string => val !== undefined)
  .forEach(val => {
    // val is string
    console.log(val.toUpperCase());
  });
```

michaelfaith commented on lib/rules/no-property-in-node.ts:
```diff
@@ -23,11 +26,14 @@ const defaultTypedNodeSourceFileTesters = [
- * @param {import('typescript').Type} type
- * @param {RegExp[]} typedNodeSourceFileTesters
+ * @param type
+ * @param typedNodeSourceFileTesters
```
Updated

michaelfaith commented on lib/rules/no-unused-message-ids.ts:
```diff
@@ -99,21 +101,27 @@ const rule = {
               hasSeenUnknownMessageId = true;
             }
-            values.forEach((val) => messageIdsUsed.add(val.value));
+            values.forEach(
+              (val) =>
+                'value' in val && messageIdsUsed.add(val.value as string),
+            );
```
That's enough to narrow `val` to Literal, but it doesn't narrow `value` to string, unfortunately. 
<img width="894" height="268" alt="image" src="https://github.com/user-attachments/assets/eb406894-d4b6-4507-ac3b-2c64f8c3d661" />

This worked, though it's a bit verbose...
```ts
values
    .filter(value => value.type === 'Literal')
    .map(value => value.value)
    .filter(value => typeof value === 'string')
    .forEach(value => messageIdsUsed.add(value));
```

michaelfaith replied:
Updated

michaelfaith commented:
I've started going through the feedback, but wanted to double check something with you. I tried to avoid changing the logic much from the original, to avoid inadvertently introducing bugs, which is why there's casting in a number of places (what I was referring to in the description).  In order to drop the casting on a lot of these, I'll need to add additional conditions and the like to narrow the types enough.  Just wanted to confirm that's the direction you want me to go?

bmish commented:
> I've started going through the feedback, but wanted to double check something with you. I tried to avoid changing the logic much from the original, to avoid inadvertently introducing bugs, which is why there's casting in a number of places (what I was referring to in the description). In order to drop the casting on a lot of these, I'll need to add additional conditions and the like to narrow the types enough. Just wanted to confirm that's the direction you want me to go?

We have pretty good tests and high test coverage, so yes I would add early returns and do other minor refactoring needed to avoid casting when practical. But this is a big PR, so if any of these improvements get too big or complicated, it would be fine to save for another time. My goal is to avoid casting as much as possible and only use it when it's a pain to avoid it.

bmish commented on tests/lib/rules/test-case-shorthand-strings.ts:
```diff
@@ -12,10 +12,10 @@ import { RuleTester } from 'eslint';
  * Returns the code for some valid test cases
- * @param {string[]} cases The code representation of valid test cases
+ * @param cases The code representation of valid test cases
  * @returns {string} Code representing the test cases
```
```suggestion
 * @returns Code representing the test cases
```

michaelfaith replied:
Updated

michaelfaith commented on lib/index.ts:
```diff
@@ -41,20 +39,56 @@ import requireMetaType from './rules/require-meta-type.js';
 
+const require = createRequire(import.meta.url);
+
+const packageMetadata = require("../package.json") as {
```
That doesn't really apply in this case, since this file's relative position to the package.json is the same in the dist folder as it is in the lib folder.

michaelfaith commented:
I believe I've addressed all of the feedback so far.  Let me know if there are any other updates you think should be made. Appreciate the thorough review.

bmish reviewed (changes requested):
It's looking good. Just a few more fixes.

Would you also mind just running some manual sanity checks on the compiled output? E.g. try loading the plugin from `dist/`. 

And make sure `dist/` is ignored by tools like the following PRs (issues I discovered during the build/publishing process):
- https://github.com/qunitjs/eslint-plugin-qunit/pull/585
- https://github.com/qunitjs/eslint-plugin-qunit/pull/602

michaelfaith commented on package.json:
```diff
@@ -75,6 +81,7 @@
     "prettier": "^3.4.1",
     "release-it": "^17.2.0",
+    "tsup": "^8.5.0",
     "typescript": "^5.8.3",
```
The version of `@typescript-eslint` that's installed doesn't include 5.9 in the peer deps yet.  I'd suggest having a separate change that updates both sets of packages.  I don't mind doing that one too.

michaelfaith commented on lib/utils.ts:
```diff
@@ -694,27 +800,33 @@ export function getSourceCodeIdentifiers(scopeManager, ast) {
   }
   return fixer.insertTextAfter(
-    sourceCode.getLastToken(node.properties.at(-1)),
+    sourceCode.getLastToken(node.properties.at(-1)!)!,
```
Updated

michaelfaith commented on tests/lib/utils.ts:
```diff
@@ -1606,12 +1822,28 @@ describe('utils', () => {
         getResult(ast) {
-          return ast.body[0].expression.right.properties[0].value.properties[0]
-            .value.properties;
+          return (
+            (
+              (
+                (
+                  (
+                    (
+                      (ast.body[0] as ExpressionStatement)
+                        .expression as AssignmentExpression
+                    ).right as ObjectExpression
+                  ).properties[0] as Property
+                ).value as ObjectExpression
+              ).properties[0] as Property
+            ).value as ObjectExpression
+          ).properties as Property[];
```
That's constructing the AST types that the tests all assume, based on the text of the string.  It was extremely tedious, using `AST Explorer` and the test strings to figure it out 😅.  There's not a great way to infer the types of what a string's AST will ultimately be parsed into, yet the tests all assume the structure of the string of the test case.  Definite opportunity for a large refactor of how these tests operate.

michaelfaith commented on lib/rules/require-meta-docs-recommended.ts:
```diff
@@ -1,31 +1,31 @@
   }
   return fixer.insertTextAfter(
-    objectNode.properties.at(-1),
+    objectNode.properties.at(-1)!,
```
Updated

bmish reviewed (approved):
Awesome work, thank you!

michaelfaith commented on .github/workflows/main.yml:
```diff
@@ -45,3 +45,13 @@ jobs:
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 'lts/*'
+      - run: npm install
```
I was thinking a bit more on this later.  In order to use `npm ci`, it'll require than just updating the calls in the workflow.  The repo would need to start maintaining a lock file.  It currently has lock files explicitly disabled, and `npm ci` requires that.  I personally don't think having a lock file is a bad thing.  I was kind of surprised when I first realized there wasn't one.  But it seemed to be a conscious choice someone made at some point.  

The other consideration is that using `npm ci` in the builds will result in longer build times, if that's important to you.
